Underscores in numeric literals.

zvezdan opened the pull request:
Since Python 3.6 underscores are allowed in numeric literals.
The underscore cannot be used at the beginning or end, nor can it be
duplicated. The change follows exactly the specification in Python
documentation referenced in the comment above it.

zvezdan committed:
Underscores in numeric literals.

Since Python 3.6 underscores are allowed in numeric literals.
The underscore cannot be used at the beginning or end, nor can it be
duplicated. The change follows exactly the specification in Python
documentation referenced in the comment above it.

zvezdan commented:
Cc: @brammool for inclusion in runtime

dpelle commented on runtime/syntax/python.vim:
```diff
@@ -157,17 +157,17 @@ syn match   pythonEscape	"\\$"
-  syn match   pythonNumber	"\<\d\+[jJ]\>"
-  syn match   pythonNumber	"\<\d\+[eE][+-]\=\d\+[jJ]\=\>"
+  " numbers (including complex)
+  syn match   pythonNumber	"\<0[oO]\%(_\=\o\)\+\>"
```
Is it intentional to only support only python3 numbers and not python2?

For example, these python2 numbers are not highlighted correctly in this PR:
```
0123  # octal, in Python3 is should 0o123
123l  # long, in Python3 is should be 123
123L  # long, in Python3 it should be 123
```
My original PR https://github.com/vim/vim/pull/8024 supported the union of python2 and python3 numbers.
I realize that python2 is end if life, but there are still plenty of python2 scripts.
Maybe we need something like `if exists("python2_highlight")` for python2 specific things?

zvezdan replied:
@dpelle Yes, this **is intentional**. I already mentioned that in the review of your PR. 🙂 

I _already_ dropped the support for Python 2 _keywords_ and _builtins_ in PR #7855.
My intent, as stated in that PR is to continue updating the syntax highlighting with the Python 3 features that are missing right now, while at the same time _dropping_ the Python 2 features. This PR is one of the steps in that process that has already been started with PR #7855.

**Rationale**: By showing the users that their code is using Python 2 syntax through the lack of syntax highlighting for those keywords/builtins or `123L` numbers, or in the next PR I'm preparing of `ur"some string"`, we will encourage the users to transition to Python 3.

The purpose of the syntax highlighting is to emphasize the correct syntax and reject/mark the incorrect.
Since the examples from Python 2 are syntactically incorrect when used in a modern Python interpreter,
```
>>> 123L
  File "<stdin>", line 1
    123L
       ^
SyntaxError: invalid syntax
```
I want to discourage everyone from using that syntax.
Python 2 is _unsupported_ for over a year now, and I intend to support only the official language features -- Python 3.

*Anyone* who needs to use Python 2 can take the copy of the syntax file *before* PR #7855 and use it locally.

zvezdan replied:
To demonstrate the intent, below is the part of my tests relevant to this change.
Notice the last section of the _bad_ examples, related to Python 2 features that are now syntactically incorrect.
```
good = (
    # From Python docs
    7, 2147483647, 0o177, 0b100110111,
    3, 79228162514264337593543950336, 0o377, 0xdeadbeef,
    100_000_000_000, 0b_1110_0101,
    3.14, 10., .001, 1e100, 3.14e-10, 0e0, 3.14_15_93,
    3.14j, 10.j, 10j, .001j, 1e100j, 3.14e-10j, 3.14_15_93j,

    # From PEP 515
    10_000_000.0, 0xCAFE_F00D, 0b_0011_1111_0100_1110, 0b_1111_0000,

    # my tests
    23, 45678, 0, 1_000_000_000_000_000, 0_0_0_0_0_0,
    0xf, 0xab123, 0XABCDEF, 0xDEADBEEF, 0x_f, 0x_FF_FF_FF_FF,
    0o17, 0o077, 0o377, 0O245, 0o456, 0o_75, 0O177_377,
    0b101, 0B1010, 0b1111111111111, 0b_1101,
    2.23, 1., 1.0, .1, 088e-08, 231_782.32_26, 231_782.32_26e3_0,
    1e3, 1.e+5, 1.0e-2, 2e0, 0e0, 1e3_2, 1e-3_2, .2_6E-1_3, 1.2_6e1_3,
    08j, 1j, 1.j, 1.0j, 1e3j, 1.e+5J, 1.0e-2j, 3.14+42.42j, 1_2.5_6e3_4j,
    -1, +3, -1.45E13J, 4.,
)

bad = (
    1.0.0, 08, 0xg, 0o8, 0b2,
    a=-1, b=+2, c=4,
    $1, j1, A.1, 1.0a, 2009.02.15,

    23_, _23, 2__3, _0, 0_, 0__0_0,
    0xf_, _0xf, 0_xf, 0x__f, 0xff__ff,
    0o7_, _0o7, 0_o7, 0o__7, 0o7__7,
    0b1_, _0b1, 0_b1, 0b__1, 0b1__1,
    1_000_000_, 1_000__000, 0o177__377, 0b_1110_,
    231__782.32, 2_.3, _2.3, 231_782.32__26, 2._3, 2.3_,
    1e_3, 1_e3, 1e3_, _1.2e3, j, _1j, _1.2j,

    # Python 2
    23L, 45678l,
    0xfL, 0XDEADBEEFL,
    0177, 0377, 0245L, 0o456L,
    0b1111111111111L,
    +4L,
)
```

brammool commented:
It should be possible to see Python2 scripts properly highlighted, I have several laying around (to be converted some day).
Is there an easy way to switch to that, after seeing some errors?

zvezdan commented:
@brammool 

I hope you had a chance to see my [reply to the review](https://github.com/vim/vim/pull/8033#discussion_r603150446) related to this topic. I already explained my rationale and reasoning there.

I see several options here:

* Use the old version of the syntax file before PR #7855 where I already dropped Python 2 keywords/builtins. This is what I mentioned in that reply above. This approach is the least burden for me as a maintainer, but I am advocating it also because this is the exact approach Python Foundation has taken with the end of support for Python 2.
* Introduce a setting in the file that would produce two effectively different syntax highlighting approaches. I could do this, but this adds an additional burden to maintain and unnecessarily complicates the code logic with too many `if` branches, for a diminishing benefit of a dwindling Python 2 code base.
* Have a separate Python 2 syntax highlight file, based on the version of my file before PR #7855 that users can instruct Vim to use when editing their old Python 2 code. Is there an easy way to do this in Vim? If it is, this is perhaps the best compromise between the first and second approach above. Setting one option to change highlighting from Python 3 to 2 (second item above) is not much different to setting another option to load a different syntax file (if possible). The latter has the benefit of keeping two separate approaches as separate files and leaving Python 2 file to slowly deprecate.

Let me know what you think and the way you want to proceed.

Considering that we already dropped syntax highlighting for Python 2 keywords/builtins, perhaps we could commit this and I would have a chance to prepare other important Python 3 updates, while also working on the compatibility layer in the way we agree on. If you prefer that we make the decision and implement the compatibility change first, that's also OK. The users who need this feature can always apply this patch directly to their syntax files until then.

Thanks!

brammool commented:
Vim has always followed the philosophy to support as many users (and uses cases) as we can, with limited extra work.  Dropping support for Python 2 syntax completely goes against that.  Those files will still exist and occasionally users will open them and look though them.  If it's riddled with error highlighting that is not nice.

I can understand that Python has moved on to Python 3 and no longer wants to spend time on old stuff.  So, what we can do is introduce the "python2" filetype.  Then we can use the previous syntax that was for Python 2 only.  It doesn't need to be maintained (except when someone spots an error).  And we don't do file type detection, the user has to manually set the filetype to "pyton2".  That way we don't need to mix two filetypes in one script.

How does that sound?

zvezdan commented:
@brammool 

> How does that sound?

That sounds fantastic. That's exactly the third item I was proposing above as the best compromise.
I just didn't know if it would be acceptable to you and if there's an easy setting in Vim (turns out that new syntax type is certainly such a setting).
I think we're on the same page here.

How do you propose we proceed?

* I could prepare a variant of my syntax file before PR #7855 with only Python 2 relevant syntax in it as _python2.vim_. I would be able to do that this weekend at the earliest. I could name it python2 and send it in a new PR, or
* We could just use the version before PR #7855 directly as _python2.vim_, but the disadvantage is that it has straddling support for Python 2/3 mixed in it at that point in time.

brammool commented:
Zvezdan Petkovic wrote:

> > How does that sound?
> 
> That sounds fantastic. That's exactly the third item I was proposing
> above as the best compromise.  I just didn't know if it would be
> acceptable to you and if there's an easy setting in Vim (turns out
> that new syntax type is certainly such a setting).
> I think we're on the same page here.
> 
> How do you propose we proceed?
> 
> * I could prepare a variant of my syntax file before PR #7855 with
> only Python 2 relevant syntax in it as _python2.vim_. I would be able
> to do that this weekend at the earliest. I could name it python2 and
> send it in a new PR, or
> * We could just use the version before PR #7855 directly as
> _python2.vim_, but the disadvantage is that it has straddling support
> for Python 2/3 mixed in it at that point in time.

It would be good to just have Python 2 support.  But I suppose the
latest Python 2 version did also include some (future) Python 3 items,
thus we should not go back to before Python 3 existed.  I think the goal
should be to keep it simple and avoid highlighting Python 2 specific
items as errors.

-- 
Bumper sticker: Honk if you love peace and quiet.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
